spark/src/main/java/org/apache/zeppelin/spark/PythonUtils.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.zeppelin.spark;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Util class for PySpark
+ */
+public class PythonUtils {
+
+  /**
+   * Get the PYTHONPATH for PySpark, either from SPARK_HOME, if it is set, or from ZEPPELIN_HOME
+   * when it is embedded mode.
+   *
+   * This method will called in zeppelin server process and spark driver process when it is
+   * local or yarn-client mode.
+   */
+  public static String sparkPythonPath() {
+    List<String> pythonPath = new ArrayList<String>();
+    String sparkHome = System.getenv("SPARK_HOME");
+    String zeppelinHome = System.getenv("ZEPPELIN_HOME");
+    if (zeppelinHome == null) {
+      zeppelinHome = new File("..").getAbsolutePath();
+    }
+    if (sparkHome != null) {
+      // non-embedded mode when SPARK_HOME is specified.
+      File pyspark = new File(sparkHome, "python/lib/pyspark.zip");
+      if (!pyspark.exists()) {
+        throw new RuntimeException("No pyspark.zip found under " + sparkHome + "/python/lib");
+      }
+      pythonPath.add(pyspark.getAbsolutePath());
+      File[] py4j = new File(sparkHome + "/python/lib").listFiles(new FilenameFilter() {
+        @Override
+        public boolean accept(File dir, String name) {
+          return name.startsWith("py4j");
+        }
+      });
+      if (py4j.length == 0) {
+        throw new RuntimeException("No py4j files found under " + sparkHome + "/python/lib");
+      } else if (py4j.length > 1) {
+        throw new RuntimeException("Multiple py4j files found under " + sparkHome + "/python/lib");
```

Comment:
One wonder if this might be a bit too restrictive ... I for one have multiple py4j versions and Spark always picks the right one

Reply:
Because spark hard code each py4j version, but zeppelin needs to work for multiple versions of spark.

Reply:
sure, but it might break people with existing setup...

Reply:
Why ? If there's multiple version of py4j under `SPARK_HOME/python/lib` or `ZEPPELIN_HOME/interpreter/spark/pyspark`, it must be something wrong with user's enviroment.

Reply:
maybe, maybe it is an edge case.
do we need to force the logic here? is it possible to call say, `sbin/spark-config.sh`?

Reply:
I think `sbin/spark-config.sh` only affect pyspark shell. Spark also hard py4j version in scala code when it needs to detect the py4j and distribute to yarn containers. . See https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/api/python/PythonUtils.scala#L35